File: sdk/swift/demo/Sources/LensCatalog.swift
import Foundation

// The asset-free post lenses the filter picker swaps between. Each one is
// a single pass over the "camera" input, activated straight from the inline
// manifest below so the demo needs no bundled lens folder for them. None
// rides the bundled reference lens instead, so it carries no inline form.
enum PostFilter: Int, CaseIterable {
    case none
    case blur
    case grade
    case bloom

    var title: String {
        switch self {
        case .none: return "None"
        case .blur: return "Blur"
        case .grade: return "Grade"
        case .bloom: return "Bloom"
        }
    }

    var manifestJSON: String? {
        switch self {
        case .none:
            return nil
        case .blur:
            return Self.manifest(id: "goss.demo.blur", name: "Blur",
                                 node: "{\"id\":\"blur\",\"type\":\"blur.pass\",\"inputs\":{\"frame\":\"camera\"},\"params\":{}}")
        case .grade:
            return Self.manifest(id: "goss.demo.grade", name: "Grade",
                                 node: "{\"id\":\"grade\",\"type\":\"grade.pass\",\"inputs\":{\"frame\":\"camera\"},\"params\":{},\"grade\":{\"exposure\":0.1,\"contrast\":1.15,\"saturation\":1.2,\"temperature\":0.05}}")
        case .bloom:
            return Self.manifest(id: "goss.demo.bloom", name: "Bloom",
                                 node: "{\"id\":\"glow\",\"type\":\"bloom.pass\",\"inputs\":{\"frame\":\"camera\"},\"params\":{},\"bloom\":{\"threshold\":0.6,\"intensity\":0.8}}")
        }
    }

    var manifestData: Data? { manifestJSON?.data(using: .utf8) }

    private static func manifest(id: String, name: String, node: String) -> String {
        "{\"glf\":\"1.0\",\"id\":\"\(id)\",\"version\":\"1.0.0\",\"display_name\":\"\(name)\",\"engine_compat\":\">=0.5\",\"capabilities\":[],\"parameters\":[],\"nodes\":[\(node)],\"triggers\":[]}"
    }
}

File: sdk/kotlin/demo/src/main/kotlin/com/gosslens/demo/LensCatalog.kt
package com.gosslens.demo

import com.gosslens.Gosslens

// The asset-free post lenses the filter picker swaps between, and the
// virtual-background composite, each an inline manifest activated straight
// from these bytes. The picker order matches filterNames below; None
// carries no manifest and deactivates the post lens entirely.
object LensCatalog {
    val filterNames = listOf("None", "Blur", "Grade", "Bloom")

    private val blur =
        """{"glf":"1.0","id":"com.gosslens.demo.blur","version":"1.0.0","display_name":"Blur","engine_compat":">=0.5","capabilities":[],"parameters":[],"nodes":[{"id":"blur","type":"blur.pass","inputs":{"frame":"camera"},"params":{}}],"triggers":[]}"""

    private val grade =
        """{"glf":"1.0","id":"com.gosslens.demo.grade","version":"1.0.0","display_name":"Grade","engine_compat":">=0.5","capabilities":[],"parameters":[],"nodes":[{"id":"grade","type":"grade.pass","inputs":{"frame":"camera"},"params":{},"grade":{"invert":1.0}}],"triggers":[]}"""

    private val bloom =
        """{"glf":"1.0","id":"com.gosslens.demo.bloom","version":"1.0.0","display_name":"Bloom","engine_compat":">=0.5","capabilities":[],"parameters":[],"nodes":[{"id":"bloom","type":"bloom.pass","inputs":{"frame":"camera"},"params":{},"bloom":{"threshold":0.6,"intensity":0.8}}],"triggers":[]}"""

    // Composites the background against the in-engine person mask, so the
    // segmenter's subject stays sharp while the background is replaced.
    private val virtualBackground =
        """{"glf":"1.0","id":"com.gosslens.demo.virtual-background","version":"1.0.0","display_name":"Virtual Background","engine_compat":">=0.5","capabilities":["segmentation"],"parameters":[],"nodes":[{"id":"background","type":"blend.pass","inputs":{"frame":"camera"},"params":{}}],"triggers":[]}"""

    /** The manifest for a filter index, or null for None (no post lens). */
    fun filterManifest(index: Int): ByteArray? = when (index) {
        1 -> blur
        2 -> grade
        3 -> bloom
        else -> null
    }?.toByteArray(Charsets.UTF_8)

    fun virtualBackgroundManifest(): ByteArray = virtualBackground.toByteArray(Charsets.UTF_8)

    // The in-engine selfie segmenter model, shipped as an app asset. When it
    // is absent at runtime the virtual-background toggle degrades rather than
    // crashing.
    const val SEGMENTER_ASSET = "selfie_segmenter.tflite"

    /** The canned gesture class the hand result carries, as a short label. */
    fun gestureName(gesture: Int): String = when (gesture) {
        Gosslens.GESTURE_CLOSED_FIST -> "fist"
        Gosslens.GESTURE_OPEN_PALM -> "open palm"
        Gosslens.GESTURE_POINTING_UP -> "point up"
        Gosslens.GESTURE_THUMB_DOWN -> "thumb down"
        Gosslens.GESTURE_THUMB_UP -> "thumb up"
        Gosslens.GESTURE_VICTORY -> "victory"
        Gosslens.GESTURE_ILOVEYOU -> "love"
        else -> "none"
    }
}
